File: src/main/java/data/SerializationManager.java
package data;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.nio.charset.StandardCharsets;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

//===========================================================//
//===========================================================//

public final class SerializationManager {
    //===========================================================//
    //===========================================================//
    // Static Field(s)

    static private final Gson s_GSON = new GsonBuilder()
        .setPrettyPrinting()
        .create();

    //===========================================================//
    //===========================================================//
    // Public Interface(s)

    static public void saveToFile(final File file, final SerializationData serData) throws IOException {
        if (!file.exists()) file.createNewFile();
        try (OutputStreamWriter writer = new OutputStreamWriter(new FileOutputStream(file), StandardCharsets.UTF_8)) 
        {
            s_GSON.toJson(serData, writer);
        }
    }

    //===========================================================//

    static public SerializationData loadFromFile(final File file) throws IOException {
        try (InputStreamReader reader = new InputStreamReader(new FileInputStream(file), StandardCharsets.UTF_8)) 
        {
            return s_GSON.fromJson(reader, SerializationData.class);
        }
    }

    //===========================================================//
    //===========================================================//
    // Constructor(s)

    private SerializationManager() {} // To prevent instantiation.
}

File: src/main/kotlin/data/SerializationManager.kt
package data

import com.google.gson.Gson
import domain.assets.security.SecurityHistory
import domain.assets.security.SecurityIdentifier
import java.io.File
import java.io.FileInputStream
import java.io.InputStreamReader
import java.nio.charset.StandardCharsets
import kotlin.time.Instant

sealed class SerializationManager {
    companion object {
        //===========================================================//
        //===========================================================//
        // Private Field(s)

        private val s_MarketDataParser = get(Type.YahooMarketDataParser)

        //===========================================================//
        //===========================================================//
        // Public Method(s)

        /**
         * Loads historical data from files based on the given parameters for backtesting algorithms.
         *
         * @param securityIdentifier the identifier of the security.
         * @param from the start date from which we want to include historical data (inclusive).
         * @param to the end date from which we don't want to include historical data (inclusive).
         * @return List of historical data entries.
         */
        fun loadHistoricalDataFromFile(securityIdentifier: SecurityIdentifier, from: Instant, to: Instant): MutableList<SecurityHistory> {
            val data = s_MarketDataParser.parse(securityIdentifier)

            return data.history
                .filter { it.date in from..to }
                .sortedBy { it.date }
                .map { SecurityHistory(it.closingPrice) }
                .toMutableList()
        }

        //===========================================================//

        internal inline fun <reified T> loadFromFile(serializer: Gson, file: File): T {
            InputStreamReader(FileInputStream(file), StandardCharsets.UTF_8).use { reader ->
                return serializer.fromJson(reader, T::class.java)
            }
        }

        //===========================================================//
        //===========================================================//
        // Private Field(s)

        private fun get(type: Type): SerializationManager {
            return when (type) {
                Type.YahooMarketDataParser -> {
                    YahooSerializedMarketDataParser
                }
            }
        }
    }

    //===========================================================//

    internal abstract fun parse(securityIdentifier: SecurityIdentifier): SecuritySerializationData

    //===========================================================//
    //===========================================================//
    // Helper Class(es)

    sealed interface Type {
        data object YahooMarketDataParser : Type
    }
}
